platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/TipsetEventCreationManager.java

```diff
@@ -0,0 +1,216 @@
+        Objects.requireNonNull(signer);
+        Objects.requireNonNull(addressBook);
+        Objects.requireNonNull(selfId);
+        Objects.requireNonNull(softwareVersion);
+        Objects.requireNonNull(transactionSupplier);
+
+        final EventCreationConfig eventCreationConfig =
+                platformContext.getConfiguration().getConfigData(EventCreationConfig.class);
+
+        eventCreator = new TipsetEventCreator(
+                platformContext, time, random, signer, addressBook, selfId, softwareVersion, transactionSupplier);
+
+        workQueue = new MultiQueueThreadConfiguration(threadManager)
+                .setThreadName("event-creator")
+                .setCapacity(eventCreationConfig.creationQueueSize())
+                .setMaxBufferSize(eventCreationConfig.creationQueueBufferSize())
+                .addHandler(EventImpl.class, this::handleEvent)
+                .addHandler(Long.class, this::handleMinimumGenerationNonAncient)
+                .setIdleCallback(this::maybeCreateEvent)
+                .setBatchHandledCallback(this::maybeCreateEvent)
+                .setWaitForWorkDuration(eventCreationConfig.creationQueueWaitForWorkPeriod())
+                .build();
+
+        final double maxCreationRate = eventCreationConfig.maxCreationRate();
+        if (maxCreationRate > 0) {
+            rateLimiter = new RateLimiter(time, eventCreationConfig.maxCreationRate());
+        } else {
+            // No brakes!
+            rateLimiter = null;
+        }
+
+        eventInserter = workQueue.getInserter(EventImpl.class);
+        minimumGenerationNonAncientInserter = workQueue.getInserter(Long.class);
+    }
+
+    /**
+     * Add an event from the event intake.
+     *
+     * @param event the event to add
+     */
+    public void registerEvent(@NonNull final EventImpl event) throws InterruptedException {
+        eventInserter.put(event);
+    }
+
+    /**
+     * Upgate the minimum generation non-ancient
+     *
+     * @param minimumGenerationNonAncient the new minimum generation non-ancient
+     */
+    public void setMinimumGenerationNonAncient(final long minimumGenerationNonAncient) throws InterruptedException {
+        minimumGenerationNonAncientInserter.put(minimumGenerationNonAncient);
+    }
+
+    /**
+     * Pass an event into the event creator.
+     *
+     * @param event the event to pass
+     */
+    private void handleEvent(@NonNull final EventImpl event) {
+        eventCreator.registerEvent(event);
+    }
+
+    /**
+     * Pass a new minimum generation non-ancient into the event creator.
+     *
+     * @param minimumGenerationNonAncient the new minimum generation non-ancient
+     */
+    private void handleMinimumGenerationNonAncient(final long minimumGenerationNonAncient) {
+        eventCreator.setMinimumGenerationNonAncient(minimumGenerationNonAncient);
+    }
+
+    /**
+     * Create a new event if it is legal to do so.
+     */
+    private void maybeCreateEvent() {
```

Comment:
This class looks like its supposed to isolate the threading, so this logic should be contained elsewhere, maybe like a separate class where the eventCreator can just call a predicate

Reply:
I've refactored this logic into the class `ThrottledTipsetEventCreator` and added unit tests. In `TipsetEventCreationManager.java`, this method is now just

```
    private void maybeCreateEvent() {
        final GossipEvent event = eventCreator.maybeCreateEvent();
        if (event != null) {
            newEventHandler.accept(event);
        }
    }
```